### PR TITLE
Trailing comma rule now encompasses function decls, exprs and types

### LINKED ISSUE #1409: Feature Request: Support trailing comma for function arguments.
Support trailing comma rule/enhance trailing comma rule for function args.

As per https://github.com/Microsoft/TypeScript/issues/7279, functions now support trailing commas for their arguments.

### PR DESCRIPTION
**Pull request by markwongsk**
Fixes #1409

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @markwongsk! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by markwongsk** [comment]
The following currently do not work:

- [ ] Typed parameters in class declarations, interface declarations, function declarations
- [ ] Methods in classes

I am working on another commit to address those cases.

**Comment by adidahiya** [comment]
cool; what's the "methods in classes" case look like?

**Comment by JKillian** [comment]
Gave it a brief glance over, looking good so far 👍

**Comment by markwongsk** [comment]
@adidahiya methods in a class or interface declaration have `node.kind` of `MethodDeclaration` and `MethodSignatureDeclaration` respectively, which aren't covered by the walker yet.

**Comment by markwongsk** [comment]
There is are a couple more issue that I have to address:
- [ ] `ObjectType` literals do not currently work
- [ ] Enums do not work

I also have a couple of test cases left. I've combed through the grammar but have no confidence on how thorough I've been. Please point out any remaining cases that you think we should support.

File: src/rules/trailingCommaRule.ts [hunk 1]
```diff
@@ -62,40 +65,186 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+    public visitTypeLiteral(node: ts.TypeLiteralNode) {
+        this.lintNode(node);
+        // object type literals need to be inspected separately because they
+        // have a different syntax list wrapper token, and they can be semicolon delimited
+        const children = node.getChildren();
+        for (let i = 0; i < children.length - 2; i++) {
+            if (children[i].kind === ts.SyntaxKind.OpenBraceToken &&
+                children[i + 1].kind === ts.SyntaxKind.SyntaxList &&
+                children[i + 2].kind === ts.SyntaxKind.CloseBraceToken) {
+/*
```
Comment (markwongsk) [inline src/rules/trailingCommaRule.ts]: need to be removed

**Comment by markwongsk** [comment]
This is now ready for review.

**Comment by JKillian** [comment]
I'm gonna test this out on our codebase, but I don't anticipate you having to change much. Do you mind actually making this PR against master? Next release is going to be TSLint 4.0 which will only support >= TS 2

**Comment by JKillian** [comment]
I think this might be a bug:

```
const colon = node.getChildren().filter((child) =>
    child.kind === ts.SyntaxKind.ColonToken //warning for missing trailing comma here
)[0];
```

**Comment by JKillian** [comment]
Requests a comma here as well - thoughts?

```
export interface TestResult {
    directory: string;
    results: {
        [fileName: string]: {
            errorsFromLinter: LintError[];
            errorsFromMarkup: LintError[];
            fixesFromLinter: string;
            fixesFromMarkup: string;
            markupFromLinter: string;
            markupFromMarkup: string;
        } // missing comma
    };
}
```

**Comment by markwongsk** [comment]
For the first one, it's a bug. For the second one, this PR makes it such that a comma is indeed necessary there (it compiles with the comma) -- similar to how if it was an object, you would need a comma there.

**Comment by JKillian** [comment]
Ah yes, you're right. Just the first one needs to get fixed, second one is fine 👍

**Comment by markwongsk** [comment]
So I took a further look into the first example, and I think the first behavior is expected after this change This is because the lambda happens to be an argument to a function, and multilined arguments to a function need to have a trailing comma. Consider
```
foo(
   arg1,
);
```
as the base behavior. It just so happens that in this case, `arg1` is a lambda. Thus,
```
foo(a =>
    doSomethingWithA(a),
);
```
I think after this change,
```
foo(
    a => doSomethingWithA(a),
);
```
might be clearer. Thoughts?

**Comment by markwongsk** [comment]
I updated the base branch to master. Let me know if you require anything else :)

**Comment by JKillian** [comment]
Looks good, thanks @markwongsk!

**Comment by timocov** [comment]
Is there workaround to skip that check? I like trailing comma in multiline "objects", but function calls it is awful. IMHO

**Pull request merged**